scipy/stats/_stats_py.py

```diff
@@ -10165,6 +10165,136 @@ def first_order(t):
+
+
+@_axis_nan_policy_factory(  # noqa: E302
+    _moment_result_object, n_samples=1, result_to_tuple=lambda x: (x,),
+    n_outputs=lambda kwds: _moment_outputs(kwds, [1, 2, 3, 4])
+)
+def lmoment(sample, order=None, *, axis=0, sorted=False, standardize=True):
+    r"""Compute L-moments of a sample from a continuous distribution
+
+    The L-moments of a probability distribution are summary statistics with
+    uses similar to those of conventional moments, but they are defined in
+    terms of the expected values of order statistics.
+    Sample L-moments are defined analogously to population L-moments, and
+    they can serve as estimators of population L-moments. They tend to be less
+    sensitive to extreme observations than conventional moments.
+
+    Parameters
+    ----------
+    sample : array_like
+        The real-valued sample whose L-moments are desired.
+    order : array_like, optional
+        The (positive integer) orders of the desired L-moments.
+        Must be a scalar or non-empty 1D array. Default is [1, 2, 3, 4].
+    axis : int, default=0
+        The axis along which to compute L-moments.
+    sorted : bool, default=False
+        Whether `sample` is already sorted in increasing order along `axis`.
+        If False (default), `sample` will be sorted.
+    standardize : bool, default=True
+        Whether to return L-moment ratios for orders 3 and higher.
+        L-moment ratios are analogous to standardized conventional
+        moments: they are the non-standardized L-moments divided
+        by the L-moment of order 2.
+
+    Returns
+    -------
+    lmoments : ndarray
+        The sample L-moments of order `order`.
+
+    See Also
+    --------
+    moment
+
+    References
+    ----------
+    .. [1] D. Bilkova. "L-Moments and TL-Moments as an Alternative Tool of
+           Statistical Data Analysis". Journal of Applied Mathematics and
+           Physics. 2014. :doi:`10.4236/jamp.2014.210104`
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> rng = np.random.default_rng(328458568356392)
+    >>> sample = rng.exponential(size=100000)
+    >>> stats.lmoment(sample)
+    array([1.00124272, 0.50111437, 0.3340092 , 0.16755338])
+
+    Note that the first four standardized population L-moments of the standard
+    exponential distribution are 1, 1/2, 1/3, and 1/6; the sample L-moments
+    provide reasonable estimates.
+
+    """
+    args = lmoment_iv(sample, order, axis, sorted, standardize)
+    sample, order, axis, sorted, standardize = args
+
+    n_moments = np.max(order)
+    k = np.arange(n_moments, dtype=sample.dtype)
+    prk = _prk(np.expand_dims(k, list(range(1, sample.ndim+1))), k)
+    bk = _br(sample, r=k)
+
+    n = sample.shape[-1]
+    bk[..., n:] = 0  # remove NaNs due to n_moments > n
+
+    lmoms = np.sum(prk * bk, axis=-1)
+    if standardize and n_moments > 2:
+        lmoms[2:] /= lmoms[1]
```

Comment:
Is that mathematically correct? For skewness and kurtosis in the context of finite precision arithmetic, I think I can present a good argument that NaN is more appropriate. SciPy and R's `moment` package agree, for instance. I'm pretty sure that even if you look at the mathematical definitions and take the the limit as deviations from a constant value approach zero, the limit is undefined.

And Hosking's `lmom` package returns NaN in these cases.
```R
library(lmom)
options(digits=16)
x = c(1, 1, 1, 1, 1)
samlmu(x, nmom=4)
```
```
l_1 l_2 t_3 t_4 
  1   0 NaN NaN 
Warning message:
In samlmu(x, nmom = 4) : all data values equal
```

Come to think of it, that's where `order = [1, 2, 3, 4]` came from - that is what Hosking chose, so it was good enough for me.

Anyway, unless we know that `0` is mathematically correct, there is evidence that `NaN` is an appropriate choice.

Reply:
Well, I suppose you could extend the definition of L-moment ratio's as "the limit of `l[r] / l[2]` as `l[2] -> 0` *from above*" 🤷🏻. 
... but know that I think about it, by this same logic, this could also be written as a 2D limit, where both `l[r]` and `l[2]` go towards 0, which isn't defined.

And in Lmo I apparantly have once decided that `inf` would be the best outcome here:

```pycon
>>> import lmo
>>> lmo.l_kurtosis(np.ones(10))
np.float64(inf)
```

I see that `scipy.stats.kurtosis` returns a `nan` in this case, but also warns about some catastrophe:

```pycon
>>> import numpy as np
>>> from scipy.stats import kurtosis
>>> kurtosis(np.ones(10))
<stdin>:1: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
np.float64(nan)
```